src/PostgREST/Config.hs

```diff
@@ -390,18 +386,52 @@ parseSecret bytes =
 
 -- | Read the JWT secret from a file if configJwtSecret is actually a filepath(has @ as its prefix).
 -- | To check if the JWT secret is provided is in fact a file path, it must be decoded as 'Text' to be processed.
-readSecretFile :: Maybe B.ByteString -> IO (Maybe B.ByteString)
-readSecretFile mSecret =
-  case (T.stripPrefix "@" . decodeUtf8) =<< mSecret of
-    Nothing       -> return Nothing
-    Just filename -> Just . chomp <$> BS.readFile (toS filename)
+readSecretFile :: AppConfig -> IO AppConfig
+readSecretFile conf =
+  case T.stripPrefix "@" . decodeUtf8 . maybe mempty identity . configJwtSecret $ conf of
+    Nothing       -> return conf
+    Just filename -> do
+      jwtSecret <- chomp <$> BS.readFile (toS filename)
+      return $ conf { configJwtSecret = Just jwtSecret }
   where
     chomp bs = fromMaybe bs (BS.stripSuffix "\n" bs)
 
+decodeSecret :: AppConfig -> IO AppConfig
+decodeSecret conf =
+  if configJwtSecretIsBase64 conf then
+    case configJwtSecret conf of
+      Just secret ->
+        case decodeB64 secret of
+          Left errMsg -> fail errMsg
+          Right bs    -> return $ conf { configJwtSecret = Just bs }
+      Nothing ->
+        return conf
+  else
+    return conf
+  where
+    decodeB64 = B64.decode . encodeUtf8 . T.strip . replaceUrlChars . decodeUtf8
+    replaceUrlChars = T.replace "_" "/" . T.replace "-" "+" . T.replace "." "="
+
+decodeJWKS :: AppConfig -> AppConfig
+decodeJWKS conf =
+  case configJwtSecret conf of
+    Nothing -> conf
+    Just secret ->
+      conf { configJWKS = Just $ parseSecret secret }
```

Comment:
```suggestion
decodeJWKS conf@AppConfig{configJwtSecret} =
  conf { configJWKS = fmap parseSecret configJwtSecret }
```

Something like that should work, too, right? fmapping over the Maybe removes a bit of the boilerplate around the maybe pattern matching.

But I don't know whether that works the same way with the other functions around here, returning `IO AppConfig`. If not, maybe it's better to leave as is for consistency.

Reply:
True, much better! All these funcs are still quite messy, I was furiously trying to get the tests back to green somehow ... :-)